pkg/ingester/metadata_map.go

```diff
@@ -0,0 +1,153 @@
+package ingester
+
+import (
+	"sync"
+	"time"
+
+	"github.com/prometheus/common/model"
+
+	"github.com/cortexproject/cortex/pkg/ingester/client"
+)
+
+type metadataMap struct {
+	mtx              sync.Mutex
+	metricToMetadata map[string]*memoryMetadataSet
+}
+
+func newMetadataMap() *metadataMap {
+	return &metadataMap{
+		metricToMetadata: map[string]*memoryMetadataSet{},
+	}
+}
+
+func (mm *metadataMap) purge(max time.Duration) int {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	var deleted int
+	now := model.Now()
+	for m, s := range mm.metricToMetadata {
+		deleted += s.purge(now, max)
+
+		if s.length() <= 0 {
+			delete(mm.metricToMetadata, m)
+		}
+	}
+
+	return deleted
+}
+
+func (mm *metadataMap) all() map[string]*memoryMetadataSet {
+	mm.mtx.Lock()
+	defer mm.mtx.Unlock()
+	metadata := mm.metricToMetadata
+	return metadata
```

Comment:
This doesn't create a copy of map (if that's what you want), and is the same as returning `mm.metricToMetadata`. Unfortunately this is broken, since now the client calling `all()` does access the same map without any lock.

It may be easier to implement forEach method (with callback function) that keeps the lock for the entire iteration.

Reply:
I removed all locks and left the locking quirks to the caller.